arch/risc-v/src/mpfs/mpfs_ihc.c

```diff
@@ -0,0 +1,1276 @@
+ *   len   - Length of the received data
+ *   src   - Source address
+ *   priv  - Handle t o private data
+ *
+ * Returned Value:
+ *   OK always
+ *
+ ****************************************************************************/
+
+static int mpfs_echo_ping_ept_cb(FAR struct rpmsg_endpoint *ept,
+                                 FAR void *data, size_t len, uint32_t src,
+                                 FAR void *priv)
+{
+  /* This is simply echoes the data back */
+
+  rpmsg_send(ept, data, len);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_echo_ping_init
+ *
+ * Description:
+ *   Initializes the rpmsg echo ping endpoint.
+ *
+ * Input Parameters:
+ *   rdev  - Rpmsg device
+ *   ept   - Endpoint data
+
+ * Returned Value:
+ *   OK on success, a negated errno otherwise
+ *
+ ****************************************************************************/
+
+static int mpfs_echo_ping_init(FAR struct rpmsg_device *rdev,
+                               FAR struct rpmsg_endpoint *ept)
+{
+  return rpmsg_create_ept(ept, rdev, MPFS_RPTUN_PING_EPT_NAME,
+                          RPMSG_ADDR_ANY, 0,
+                          mpfs_echo_ping_ept_cb, NULL);
+}
+
+/****************************************************************************
+ * Name: mpfs_rpmsg_device_created
+ *
+ * Description:
+ *   Callback that is called when the underlying rpmsg device has been
+ *   created. This is used to initialize the ping enpoint at the proper
+ *   time.
+ *
+ * Input Parameters:
+ *   rdev   - Rpmsg device
+ *   priv_  - Private data
+
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_rpmsg_device_created(FAR struct rpmsg_device *rdev,
+                                      FAR void *priv_)
+{
+  struct rpmsg_virtio_device *vdev = container_of(rdev,
+                                                  struct rpmsg_virtio_device,
+                                                  rdev);
+
+  g_mpfs_virtio_device = vdev;
+  g_mpfs_rpmsg_device  = rdev;
+
+  g_mpfs_virtqueue_table[0].data = g_mpfs_virtio_device->svq;
+  g_mpfs_virtqueue_table[1].data = g_mpfs_virtio_device->rvq;
+
+  mpfs_echo_ping_init(rdev, &g_mpgs_echo_ping_ept);
+}
+
+/****************************************************************************
+ * Name: mpfs_rptun_thread
+ *
+ * Description:
+ *   This is used to listen to the g_mpfs_rx_lock semaphore and then
+ *   notifying the associated virtqueue.  The virtqueue_notification()
+ *   cannot be called from irq context.
+ *
+ * Input Parameters:
+ *   argc   - Argument count
+ *   argv   - Argument variables
+
+ * Returned Value:
+ *   0 on exit
+ *
+ ****************************************************************************/
+
+static int mpfs_rptun_thread(int argc, FAR char *argv[])
+{
+  struct mpfs_queue_table *info;
+
+  while (1)
+    {
+      nxsem_wait(&g_mpfs_rx_lock);
```

Comment:
based on a usage I think it is better to name it `g_mpfs_rx_sig` and not `g_mpfs_rx_lock`